### PR TITLE
VAULT-6368 Metrics-only listener for Agent

### PR DESCRIPTION
**Pull request by VioletHynes**
This introduces a new option for the Agent HCL config, `metrics_only_listener`. It defaults to false. If set to true, it will make the defined listener only serve metrics. It can be put alongside other listener blocks, too, like so:

```
cache {
 enforce_consistency = "always"
 use_auto_auth_token = true
}

listener "tcp" {
    address = "127.0.0.1:8100"
    tls_disable = true
}

listener "tcp" {
    address = "127.0.0.1:3000"
    tls_disable = true
    telemetry {
      metrics_only_listener = true
    }
}
```

Using the above, you can e.g. cURL the metrics (but nothing else). This shows an example with an auth failure:
```
curl http://127.0.0.1:3000/agent/v1/metrics
{"Timestamp":"2022-11-23 17:41:50 +0000 UTC","Gauges":[{"Name":"vault.runtime.alloc_bytes","Value":10344560,"Labels":{}},{"Name":"vault.runtime.free_count","Value":52367,"Labels":{}},{"Name":"vault.runtime.heap_objects","Value":45838,"Labels":{}},{"Name":"vault.runtime.malloc_count","Value":98205,"Labels":{}},{"Name":"vault.runtime.num_goroutines","Value":13,"Labels":{}},{"Name":"vault.runtime.sys_bytes","Value":30328072,"Labels":{}},{"Name":"vault.runtime.total_gc_pause_ns","Value":375709,"Labels":{}},{"Name":"vault.runtime.total_gc_runs","Value":5,"Labels":{}}],"Points":[],"Counters":[{"Name":"vault.agent.auth.failure","Count":1,"Rate":0.1,"Sum":1,"Min":1,"Max":1,"Mean":1,"Stddev":0,"Labels":{}}],"Samples":[]}
```

Note that we still do need a cache block for this listener, though we have a separate work item tracked to enable listeners without a cache, and I felt it best to keep the two work items distinct.

### CODE REVIEW

File: internalshared/configutil/config.go [hunk 1]
```diff
@@ -47,25 +48,6 @@ type SharedConfig struct {
 	ClusterName string `hcl:"cluster_name"`
 }
 
-// LoadConfigFile loads the configuration from the given file.
```
Comment (VioletHynes) [inline internalshared/configutil/config.go:50]: These two functions were not used anywhere, so I cleaned them up.

**Comment by ncabatoff** [comment]
I have other things to get to before I review this, but just one question based on the description: why do we need `metrics_only_listener`?  Can't we rely solely on the presence of a `telemetry` block?  I just worry that otherwise it may seem that you can have a telemetry block with metrics_only_listener=false.

Alternatively, maybe the top-level block should be named `metrics_listener` instead of `listener`?

Also, can you look at the telemetry block in Vault server config and see what we should adopt from there?

**Comment by VioletHynes** [comment]
I had originally looked at potentially creating a `metrics_listener` construct, but as I worked through it, it did not seem to make less and less sense. The listener blocks you can define for Vault Agent use the same shared config as Vault itself does. Keeping it this way means that you can utilise all of the same options as you can with Vault Server `listener` blocks.

So, to the point of:
`Also, can you look at the telemetry block in Vault server config and see what we should adopt from there?`
There is one:
`unauthenticated_metrics_access`
You can specify it and it parses fine (since we share the listener blocks), but it doesn't actually do anything for Agent.

I'd rather not diverge the Agent listener config from the Vault listener config, so I'd rather the presence of a telemetry block did not have special behaviour for Agent that it doesn't for Vault, for example. I also think we gain a lot from sharing the listener config.

**Comment by ncabatoff** [comment]
Ok.  What about the non-listener telemetry stanza, i.e. https://developer.hashicorp.com/vault/docs/configuration/telemetry ?

**Comment by VioletHynes** [comment]
I'm a little confused as to what you mean - we can't rely on items inside the telemetry stanza to determine how listeners should behave. Agent supports the telemetry stanza in the same way as Server - if we're missing functionality, that would be news to me.

See also: https://developer.hashicorp.com/vault/docs/agent#telemetry-stanza

Perhaps I misunderstood what you meant, though?

**Comment by ncabatoff** [comment]
> I'm a little confused as to what you mean - we can't rely on items inside the telemetry stanza to determine how listeners should behave. Agent supports the telemetry stanza in the same way as Server - if we're missing functionality, that would be news to me.
> 
> See also: https://developer.hashicorp.com/vault/docs/agent#telemetry-stanza
> 
> Perhaps I misunderstood what you meant, though?

Nope, I just forgot that we'd already added some metrics support to Agent.  Nevermind!

**Comment by ncabatoff** [comment]
Another high-level thought: rather than a bool, how about introducing a string field?  So instead of metrics_only_listener, make it a top-level listener field e.g. `listener_function` which can have a value of either `telemetry` or `caching-proxy` - or possible also `proxy`, which might help us address https://github.com/hashicorp/vault/issues/8953.

**Comment by peteski22** [comment]
> `listener_function`

Sorry, late to the party, is this to imply the `role` the listener would be performing? 

Would this look something like this:

```
listener "tcp" {
    address = "127.0.0.1:3000"
    listener_function = "telemetry"
}
```

**Comment by VioletHynes** [comment]
Ideally I'd like this functionality to be potentially-extendable to Vault-actual if possible. While I don't know of any requests for this feature, the listener block is shared between the two and I'd rather not have separate configs in Agent/Vault as a result.

Part of the reason I'm in favour of a boolean over a string is that I see 'metrics only listener' is a very common use case and pattern that (in my mind) is distinct to other software-specifc configuration options. In other words, I definitely see this kind of thing as a 'metrics only mode' that you can opt into for security reasons as opposed to something we'll apply to other APIs.

As a result, I do have a preference for the bool, as I feel like a lot of the string options read awkwardly, but I don't feel massively strongly about it and open to changing.

I tried to come up with a nice sounding string option, maybe something like these?:
```
apis_served = "all"
apis_served = "metrics"
```

```
listener_function = "normal"
listener_function = "metrics"
```

```
listener_type = "regular"
listener_type = "metrics"
```

Looking for additional thoughts!

**Comment by peteski22** [comment]
Could you not just have `default` and `metrics` (where `default` is also what we get when the field isn't specified)? 

Really sorry if I'm being dumb.

*Metrics Listener*:
```
listener "tcp" {
    address = "127.0.0.1:3000"
    role = "metrics"
}
```

*Business as usual*:
```
listener "tcp" {
    address = "127.0.0.1:3000"
}
```

or

```
listener "tcp" {
    address = "127.0.0.1:3000"
    role = "default"
}
```

I'm using `role` as it's the role that listener is going to perform, but also we don't seem to use `listener_address` for example. So it feels (to me) like we should pick something from the second part of the names (`role`, `type`, `function`).

I guess using a `string` vs `bool` type does give you flexibility to add other roles later. That is what Nick was suggesting, that we could add ones to resolve an issue where one config stanza (`listener`) is ignored unless some other stanza (`cache`) is also provided?

**Comment by VioletHynes** [comment]
I think this issue is separate from the issue where the `listener` stanza is currently linked to the `cache` stanza. I mentioned it briefly in the description, but I don't think that this really ties into that work at all (if we do enable `listener` without `cache`, we should probably just use the cache if we have it and don't if we don't). 

I opted not to try and tackle the listener/cache entanglement here to keep those pieces of work distinct, but when we do that work, I think it would be a bad idea to require a new type of listener for it - we should support `listeners` without the cache without any other special configuration. Customers broadly just want it to work without needing to specify a `cache`.

The above approach of having a `role` be something like `caching-proxy` also makes this option incompatible with Vault's `listener` stanza, which I don't particularly like, and makes it hard to introduce `roles` for Vault listeners (we'd need to completely disentangle the code). 

We can't use `type`, for the name, as that's taken, and I guess `role` works, but I'm currently a little unconvinced that there are is a use case for having more `roles` to begin with, and that whatever those roles are are incompatible with the other listener `roles`. To me, it also makes it a little less immediately clear to me that this turns off all of the other APIs, but I guess we could use `metrics_only` or similar to make it clearer. 

If you both think that the string approach is the right approach, and we think `role` and `default`/`metrics_only` is clear enough for both Vault and Vault Agent, I'm happy to do it!

File: internalshared/configutil/listener.go [hunk 2]
```diff
@@ -22,6 +22,8 @@ type ListenerTelemetry struct {
 	UnusedKeys                      UnusedKeyMap `hcl:",unusedKeyPositions"`
 	UnauthenticatedMetricsAccess    bool         `hcl:"-"`
 	UnauthenticatedMetricsAccessRaw interface{}  `hcl:"unauthenticated_metrics_access,alias:UnauthenticatedMetricsAccess"`
+	MetricsOnlyListener             bool         `hcl:"-"`
+	MetricsOnlyListenerAccessRaw    interface{}  `hcl:"metrics_only_listener,alias:MetricsOnlyListener"`
```
Comment (ncabatoff) [inline internalshared/configutil/listener.go:26]:
```suggestion
	MetricsOnlyListenerRaw    interface{}  `hcl:"metrics_only_listener,alias:MetricsOnlyListener"`
```

File: internalshared/configutil/listener.go [hunk 3]
```diff
@@ -348,6 +350,14 @@ func ParseListeners(result *SharedConfig, list *ast.ObjectList) error {
 
 				l.Telemetry.UnauthenticatedMetricsAccessRaw = nil
 			}
+
+			if l.Telemetry.MetricsOnlyListenerAccessRaw != nil {
+				if l.Telemetry.MetricsOnlyListener, err = parseutil.ParseBool(l.Telemetry.MetricsOnlyListenerAccessRaw); err != nil {
+					return multierror.Prefix(fmt.Errorf("invalid value for telemetry.metrcs_only_listener: %w", err), fmt.Sprintf("listeners.%d", i))
```
Comment (ncabatoff) [inline internalshared/configutil/listener.go:356]:
```suggestion
					return multierror.Prefix(fmt.Errorf("invalid value for telemetry.metrics_only_listener: %w", err), fmt.Sprintf("listeners.%d", i))
```

**Comment by VioletHynes** [comment]
Updated to use `role`, with `default` and `metrics_only` options! Thanks for the comments.

File: website/content/docs/agent/caching/index.mdx [hunk 4]
```diff
@@ -227,7 +227,9 @@ These are common configuration values that live within the `persist` block:
 There can be one or more `listener` blocks at the top level. These configuration
 values are common to both `tcp` and `unix` listener blocks. Blocks of type
 `tcp` support the standard `tcp` [listener](/docs/configuration/listener/tcp)
-options.
+options. Additionally, the `role` string option is available as part of the top level
+of the `listener` block, which can be configured to `metrics_only` to serve only metrics,
+the default role, `default`, serves everything (including metrics).
```
Comment (ncabatoff) [inline website/content/docs/agent/caching/index.mdx:232]:
```suggestion
or the default role, `default`, which serves everything (including metrics).
```

File: command/agent.go [hunk 5]
```diff
@@ -700,18 +700,20 @@ func (c *AgentCommand) Run(args []string) int {
 			// Parse 'require_request_header' listener config option, and wrap
 			// the request handler if necessary
 			muxHandler := cacheHandler
-			if lnConfig.RequireRequestHeader {
+			if lnConfig.RequireRequestHeader && !("metrics_only" == lnConfig.Role) {
```
Comment (ncabatoff) [inline command/agent.go:703]: Curious why you aren't using `!=` here and below?

**Review by ncabatoff — APPROVED** [review]

File: command/agent.go (hunk 5 above)
Comment (VioletHynes) [reply command/agent.go:703]: I think this is my years of Java usage seeping into my tendencies, and you're right, I should change this to be clearer

In Java, the 'correct' way to compare string-not-equals is:
`!"const_string".equals(stringObject)`
You do the const string first, as `stringObject` could be `null`

I'll update this to use `!=` instead, as it's easier at a glance to understand. But it's funny, thinking about the reason I'd made that decision. I didn't even realise, at first.

**Pull request merged**